wpilib/src/i2c.rs

```diff
@@ -0,0 +1,197 @@
+    ///   transaction is complete.
+    ///
+    /// Returns a result based on whether the transaction was successful
+    pub fn write_bulk(&self, data: &[u8]) -> io::Result<usize> {
+        let status = unsafe {
+            HAL_WriteI2C(
+                self.port as HAL_I2CPort::Type,
+                i32::from(self.device_address),
+                data.as_ptr(),
+                data.len() as i32,
+            )
+        };
+
+        io_result(status)
+    }
+
+    /// Execute a read transaction with the device.
+    ///
+    /// Read bytes from a device.
+    /// Most I2C devices will auto-increment the register pointer internally
+    /// allowing you to read consecutive registers on a device in a single
+    /// transaction.
+    ///
+    /// Returns a result based on whether the transaction was successful
+    pub fn read(&self, register_address: i32, buf: &mut [u8]) -> io::Result<usize> {
+        if buf.is_empty() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "Write buffer length < 1",
+            ));
+        }
+
+        self.transaction(&[register_address as u8], buf)
+    }
+
+    /// Execute a read only transaction with the device.
+    ///
+    /// Read bytes from a device. This method does not write any data to prompt the
+    /// device.
+    ///
+    /// Returns a result based on whether the transaction was successful
+    pub fn read_only(&self, buf: &mut [u8]) -> io::Result<usize> {
+        let status = unsafe {
+            HAL_ReadI2C(
+                self.port as HAL_I2CPort::Type,
+                i32::from(self.device_address),
+                buf.as_mut_ptr(),
+                buf.len() as i32,
+            )
+        };
+
+        io_result(status)
+    }
+
+    /// Verify that a device's registers contain expected values.
+    ///
+    /// Most devices will have a set of registers that contain a known value that
+    /// can be used to identify them.  This allows an I2C device driver to easily
+    /// verify that the device contains the expected value.
+    ///
+    /// The device must support and be configured to use register
+    /// auto-increment.
+    pub fn verify_sensor(&self, register_address: i32, expected: &[u8]) -> bool {
+        let mut i = 0;
+        let mut cur_register_address = register_address;
+
+        loop {
+            if i >= expected.len() {
+                break;
+            }
+
+            let to_read = if expected.len() - i < 4 {
+                expected.len() - i
+            } else {
+                4
+            };
+
+            let mut buf = vec![0; to_read];
+
+            if self.read(cur_register_address, &mut buf[..]).is_err() {
+                return false;
+            }
+
+            for j in 0..to_read {
+                if buf[j] != expected[i + j] {
+                    return false;
+                }
+            }
+
+            i += 4;
+            cur_register_address += 4;
```

Comment:
Surely there's a more idiomatic way to iterate by groups of 4 without having to directly index.